decoder/pokestop_showcase.go

```diff
@@ -0,0 +1,157 @@
+}
+type contestPokemonTypeFocus struct {
+	*pogo.ContestPokemonTypeFocusProto
+}
+type contestPokemonAlignmentFocus struct {
+	*pogo.ContestPokemonAlignmentFocusProto
+}
+type contestPokemonClassFocus struct {
+	*pogo.ContestPokemonClassFocusProto
+}
+type contestPokemonFamilyFocus struct {
+	*pogo.ContestPokemonFamilyFocusProto
+}
+type contestBuddyFocus struct {
+	*pogo.ContestBuddyFocusProto
+}
+type contestGenerationFocus struct {
+	*pogo.ContestGenerationFocusProto
+}
+type contestHatchedFocus struct {
+	*pogo.ContestHatchedFocusProto
+}
+type contestMegaFocus struct {
+	*pogo.ContestTemporaryEvolutionFocusProto
+}
+type contestShinyFocus struct {
+	*pogo.ContestShinyFocusProto
+}
+
+func (p *contestPokemonFocus) Requirements() map[string]any {
+	result := make(map[string]any)
+	result["pokemon_id"] = p.PokedexId
+	if p.RequireFormToMatch {
+		result["pokemon_form"] = p.PokemonDisplay.Form
+	}
+	return result
+
+}
+func (p *contestPokemonTypeFocus) Requirements() map[string]any {
+	result := make(map[string]any)
+	result["pokemon_type_1"] = p.GetPokemonType1()
+	if type2 := p.GetPokemonType2(); type2 != pogo.HoloPokemonType_POKEMON_TYPE_NONE {
+		result["pokemon_type_2"] = type2
+	}
+	return result
+}
+func (p *contestPokemonAlignmentFocus) Requirements() map[string]any {
+	// unset, purified, shadow
+	return map[string]any{
+		"pokemon_alignment": p.GetRequiredAlignment(),
+	}
+}
+func (p *contestPokemonClassFocus) Requirements() map[string]any {
+	// normal, legendary, mythic, ultra beast
+	return map[string]any{
+		"pokemon_class": p.GetRequiredClass(),
+	}
+}
+func (p *contestPokemonFamilyFocus) Requirements() map[string]any {
+	// family pikachu, zubat e.g.
+	return map[string]any{
+		"pokemon_family": p.GetRequiredFamily(),
+	}
+}
+func (p *contestBuddyFocus) Requirements() map[string]any {
+	return map[string]any{
+		"min_level": p.GetMinBuddyLevel(),
+	}
+}
+func (p *contestGenerationFocus) Requirements() map[string]any {
+	// gen 1 - 9
+	return map[string]any{
+		"generation": p.GetPokemonGeneration(),
+	}
+}
+func (p *contestHatchedFocus) Requirements() map[string]any {
+	return map[string]any{
+		"hatched": p.GetRequireToBeHatched(),
+	}
+}
+func (p *contestMegaFocus) Requirements() map[string]any {
+	// GetRestriction()                 -> MEGA, NOT_TEMP_EVO
+	// GetTemporaryEvolutionRequired()  -> MEGA, MEGA_X, MEGA_Y, PRIMAL
+	return map[string]any{
+		"temp_evolution": p.GetTemporaryEvolutionRequired(),
+		"restriction":    p.GetRestriction(),
+	}
+}
+func (p *contestShinyFocus) Requirements() map[string]any {
+	return map[string]any{
+		"shiny": p.GetRequireToBeShiny(),
+	}
+}
+
+func createFocusStoreFromContestProto(contest *pogo.ContestProto) map[contestFocusType]contestFocusRequirementProvider {
+	focusStore := make(map[contestFocusType]contestFocusRequirementProvider)
+
+	for _, focus := range contest.GetFocuses() {
+		if pok := focus.GetPokemon(); pok != nil {
+			focusStore[focusPokemon] = &contestPokemonFocus{pok}
```

Comment:
interfaces are a pointer anyway, can't you just assign this without the address?
I'm also wondering why you don't just create the whole map here rather than  even use interfaces at all? It seems like you return this structure, then immediately call Requirements on it so why not just return an appropriate fully populated map
eg `map[contestFocusType]map[string]any` ?
and then you don't even need any interfaces you just do the extraction in this function so it is simpler to understand.
Perhaps I am missing the purpose of the extra data structures and all the interfaces?

Reply:
Hmm you might be right, probably just thought too complicated about it initially, will refactor that bit

Reply:
If you can't have more than one focus, it may as well just return the final map[string]any (but I think you can have more than one focus?)

Reply:
For now it's just the pokemon focus which appears twice if they use form.... Then there is one focus with pokemon without form and one with the normal form, like it happened for normal/galarian ponyta... Only the galarian ponyta had one focus, with the galarian form... :)